File: impl/go/value.go
package twill

import "strings"

// ValueNode is a node of a parsed declaration value.
type ValueNode interface {
	isValueNode()
}

// ValueWord is a word.
type ValueWord struct{ Value string }

// ValueSeparator is `,`, `/`, or whitespace.
type ValueSeparator struct{ Value string }

// ValueFunction is a function call; Name is empty for a bare `( ... )`.
type ValueFunction struct {
	Name  string
	Nodes []ValueNode
}

func (*ValueWord) isValueNode()      {}
func (*ValueSeparator) isValueNode() {}
func (*ValueFunction) isValueNode()  {}

// Word creates a word node.
func Word(v string) *ValueWord { return &ValueWord{Value: v} }

// Sep creates a separator node.
func Sep(v string) *ValueSeparator { return &ValueSeparator{Value: v} }

// Fn creates a function node.
func Fn(name string, nodes ...ValueNode) *ValueFunction {
	if nodes == nil {
		nodes = []ValueNode{}
	}
	return &ValueFunction{Name: name, Nodes: nodes}
}

// ParseValue parses a value into words, separators, and function calls.
func ParseValue(input string) []ValueNode {
	ast := []ValueNode{}
	var stack []*ValueFunction
	var parent *ValueFunction
	var buffer strings.Builder

	push := func(n ValueNode) {
		if parent != nil {
			parent.Nodes = append(parent.Nodes, n)
		} else {
			ast = append(ast, n)
		}
	}
	flush := func() {
		if buffer.Len() > 0 {
			push(Word(buffer.String()))
			buffer.Reset()
		}
	}

	for i := 0; i < len(input); i++ {
		c := input[i]
		switch {
		case c == '\\':
			end := i + 2
			if end > len(input) {
				end = len(input)
			}
			buffer.WriteString(input[i:end])
			i++
		case c == '"' || c == '\'':
			start := i
			for i++; i < len(input); i++ {
				d := input[i]
				if d == '\\' {
					i++
				} else if d == c {
					break
				}
			}
			end := i + 1
			if end > len(input) {
				end = len(input)
			}
			buffer.WriteString(input[start:end])
		case c == '(':
			node := Fn(buffer.String())
			buffer.Reset()
			push(node)
			stack = append(stack, parent)
			parent = node
		case c == ')':
			flush()
			if parent != nil {
				parent = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
			} else {
				buffer.WriteByte(')')
			}
		case c == ',' || c == '/':
			flush()
			push(Sep(string(c)))
		case isSpace(c):
			flush()
			j := i
			for j < len(input) && isSpace(input[j]) {
				j++
			}
			push(Sep(input[i:j]))
			i = j - 1
		default:
			buffer.WriteByte(c)
		}
	}
	flush()
	return ast
}

// ValueToCSS prints value nodes back to CSS text.
func ValueToCSS(nodes []ValueNode) string {
	var b strings.Builder
	valueToCSS(&b, nodes)
	return b.String()
}

func valueToCSS(b *strings.Builder, nodes []ValueNode) {
	for _, n := range nodes {
		switch n := n.(type) {
		case *ValueWord:
			b.WriteString(n.Value)
		case *ValueSeparator:
			b.WriteString(n.Value)
		case *ValueFunction:
			b.WriteString(n.Name)
			b.WriteByte('(')
			valueToCSS(b, n.Nodes)
			b.WriteByte(')')
		}
	}
}

// ValueVisitor visits a value node. It may return replacement nodes (which
// are not revisited) and whether to skip the node's children.
type ValueVisitor func(n ValueNode, parent *ValueFunction) (replacement []ValueNode, replaced bool, skip bool)

// WalkValue walks value nodes depth first with replacement support.
func WalkValue(nodes *[]ValueNode, visit ValueVisitor) {
	walkValue(nodes, visit, nil)
}

func walkValue(nodes *[]ValueNode, visit ValueVisitor, parent *ValueFunction) {
	for i := 0; i < len(*nodes); i++ {
		n := (*nodes)[i]
		replacement, replaced, skip := visit(n, parent)
		if replaced {
			rest := append([]ValueNode{}, (*nodes)[i+1:]...)
			*nodes = append(append((*nodes)[:i], replacement...), rest...)
			i += len(replacement) - 1
			continue
		}
		if skip {
			continue
		}
		if f, ok := n.(*ValueFunction); ok {
			walkValue(&f.Nodes, visit, f)
		}
	}
}

File: impl/ocaml/src/value.ml
(* Declaration value parsing (SPEC §4.1.8). *)

type vnode = Word of word | Sep of string | Fn of fn
and word = { mutable w : string }
and fn = { mutable fname : string; mutable fnodes : vnode list }

let word s = Word { w = s }
let sep s = Sep s
let fn name nodes = Fn { fname = name; fnodes = nodes }

(* Parses a value into words, separators, and function calls. *)
let parse input : vnode list =
  let n = String.length input in
  let ast = ref [] in
  let stack = ref [] in
  let parent : fn option ref = ref None in
  let buffer = Buffer.create 32 in
  let push node =
    match !parent with Some f -> f.fnodes <- f.fnodes @ [ node ] | None -> ast := !ast @ [ node ]
  in
  let flush () =
    if Buffer.length buffer > 0 then begin
      push (word (Buffer.contents buffer));
      Buffer.clear buffer
    end
  in
  let i = ref 0 in
  while !i < n do
    let c = input.[!i] in
    if c = '\\' then begin
      let end_ = min (!i + 2) n in
      Buffer.add_string buffer (String.sub input !i (end_ - !i));
      i := !i + 2
    end
    else if c = '"' || c = '\'' then begin
      let start = !i in
      let j = ref (!i + 1) in
      let closed = ref false in
      while (not !closed) && !j < n do
        let d = input.[!j] in
        if d = '\\' then j := !j + 2 else if d = c then closed := true else incr j
      done;
      let end_ = min (!j + 1) n in
      Buffer.add_string buffer (String.sub input start (end_ - start));
      i := end_
    end
    else if c = '(' then begin
      let node = { fname = Buffer.contents buffer; fnodes = [] } in
      Buffer.clear buffer;
      push (Fn node);
      stack := !parent :: !stack;
      parent := Some node;
      incr i
    end
    else if c = ')' then begin
      flush ();
      (match !parent with
      | Some _ -> (
          match !stack with
          | p :: rest -> parent := p; stack := rest
          | [] -> parent := None)
      | None -> Buffer.add_char buffer ')');
      incr i
    end
    else if c = ',' || c = '/' then begin
      flush ();
      push (sep (String.make 1 c));
      incr i
    end
    else if Ast.is_space c then begin
      flush ();
      let j = ref !i in
      while !j < n && Ast.is_space input.[!j] do incr j done;
      push (sep (String.sub input !i (!j - !i)));
      i := !j
    end
    else (Buffer.add_char buffer c; incr i)
  done;
  flush ();
  !ast

let rec to_css_buf b nodes =
  List.iter
    (function
      | Word w -> Buffer.add_string b w.w
      | Sep s -> Buffer.add_string b s
      | Fn f ->
          Buffer.add_string b f.fname;
          Buffer.add_char b '(';
          to_css_buf b f.fnodes;
          Buffer.add_char b ')')
    nodes

(* Prints value nodes back to CSS text. *)
let to_css nodes =
  let b = Buffer.create 64 in
  to_css_buf b nodes;
  Buffer.contents b

type vaction = VContinue | VSkip | VReplace of vnode list

(* Walks value nodes depth first; replacements are not revisited. *)
let rec walk_from (nodes : vnode list ref) visit (parent : fn option) =
  let rec loop before rest =
    match rest with
    | [] -> ()
    | node :: tail -> (
        match visit node parent with
        | VReplace repl ->
            nodes := List.rev_append before (repl @ tail);
            loop (List.rev_append repl before) tail
        | VSkip -> loop (node :: before) tail
        | VContinue ->
            (match node with
            | Fn f ->
                let r = ref f.fnodes in
                walk_from r visit (Some f);
                f.fnodes <- !r
            | _ -> ());
            loop (node :: before) tail)
  in
  loop [] !nodes

let walk nodes visit = walk_from nodes visit None
